pkg/wasm/cache.go

```diff
@@ -100,52 +101,63 @@ func NewLocalFileCache(dir string, purgeInterval, moduleExpiry time.Duration) *L
 
 // Get returns path the local Wasm module file.
 func (c *LocalFileCache) Get(downloadURL, checksum string, timeout time.Duration) (string, error) {
-	url, err := url.Parse(downloadURL)
-	if err != nil {
-		return "", fmt.Errorf("fail to parse Wasm module fetch url: %s", downloadURL)
-	}
 	// Construct Wasm cache key with downloading URL and provided checksum of the module.
 	key := cacheKey{
 		downloadURL: downloadURL,
 		checksum:    checksum,
 	}
 
-	switch url.Scheme {
-	case "http", "https":
-		// First check if the cache entry is already downloaded.
-		if modulePath := c.getEntry(key); modulePath != "" {
-			return modulePath, nil
-		}
+	// First check if the cache entry is already downloaded.
+	if modulePath := c.getEntry(key); modulePath != "" {
+		return modulePath, nil
+	}
+
+	u, err := url.Parse(downloadURL)
+	if err != nil {
+		return "", fmt.Errorf("fail to parse Wasm module fetch url: %s", downloadURL)
+	}
 
-		// If the module is not available locally, download the Wasm module with http fetcher.
-		b, err := c.httpFetcher.Fetch(downloadURL, timeout)
+	// If not, fetch images
+	var b []byte
+	switch u.Scheme {
+	case "http", "https":
+		// Download the Wasm module with http fetcher.
+		b, err = c.httpFetcher.Fetch(downloadURL, timeout)
 		if err != nil {
 			wasmRemoteFetchCount.With(resultTag.Value(downloadFailure)).Increment()
 			return "", err
 		}
-
-		// Get sha256 checksum and check if it is the same as provided one.
-		dChecksum := fmt.Sprintf("%x", sha256.Sum256(b))
-		if checksum != "" && dChecksum != checksum {
-			wasmRemoteFetchCount.With(resultTag.Value(checksumMismatch)).Increment()
-			return "", fmt.Errorf("module downloaded from %v has checksum %v, which does not match: %v", downloadURL, dChecksum, checksum)
+	case "oci":
+		ctx, cancel := context.WithTimeout(context.Background(), timeout)
+		defer cancel()
+		// TODO: support imagePullSecret and pass it to ImageFetcherOption.
+		fetcher := NewImageFetcher(ctx, ImageFetcherOption{})
+		b, err = fetcher.Fetch(u.Host + u.Path)
+		if err != nil {
+			return "", fmt.Errorf("could not fetch Wasm OCI image: %v", err)
 		}
+	default:
+		return "", fmt.Errorf("unsupported Wasm module downloading URL scheme: %v", u.Scheme)
+	}
 
-		wasmRemoteFetchCount.With(resultTag.Value(fetchSuccess)).Increment()
+	// Get sha256 checksum and check if it is the same as provided one.
```

Comment:
the API docs currently state that the sha256 field will be used to verify the *container checksum* when using OCI images. we should adhere to that here (or ignore it for now)

Reply:
oh is that? what exactly does it mean by container checksum? Manifest's digest? or layer?

Reply:
container digest == manifest digest afaik

Reply:
Should we only do checksum checking for http fetching, and verify fetched image digest instead for OCI?

Reply:
done https://github.com/istio/istio/pull/34430/commits/94714d1908b40ee2d47fea3201f42a90758dc5c2